lib/ret_web/channels/hub_channel.ex

```diff
@@ -1445,14 +1443,49 @@ defmodule RetWeb.HubChannel do
     %{nid: nid, root_nid: root_nid, update_message: Jason.encode!(update_message)}
   end
 
-  defp parse(%{"nid" => nid, "create_message" => create_message, "updates" => updates}) do
+  defp parse(
+         %{
+           "nid" => nid,
+           "create_message" => create_message,
+           "updates" => updates
+         } = params
+       ) do
     %{
       nid: nid,
       create_message: Jason.encode!(create_message),
-      updates: Enum.map(updates, &parse/1)
+      updates: Enum.map(updates, &parse/1),
+      promotion_token: Map.get(params, "promotion_token", nil),
+      file_id: Map.get(params, "file_id", nil),
+      file_access_token: Map.get(params, "file_access_token", nil)
     }
   end
 
+  defp handle_file_promotion(entity, %{file_id: nil} = _params, _account, socket) do
+    broadcast_entity_state(entity, socket)
+  end
```

Comment:
My guess is that this pattern is not idiomatic elixir, because your intent is something like "If there's a `file_id`, try to promote/activate it (potentially returning an error), then broadcast. If there's not a `file_id`, just broadcast." And it feels strange to stick the "then broadcast" part inside `handle_file_promotion`. 

But I don't know elixir well enough to know how experts tend to write this (common) case. No need to change it, but if you find a better way, please let me know because I'll be curious to learn it. 

One idea is that `maybe_promote_file` returns `:ok` or `{:error, reason}`, and the `file_id: nil` case is just `:ok`. Then the calling code uses `if` or `with`.

Reply:
Ah yes, I should've noticed an anti-pattern forming when I found myself chaining responsibility there. Took your suggestion and like it much better!